src/main/java/org/truffleruby/core/thread/TruffleThreadNodes.java

```diff
@@ -0,0 +1,60 @@
+package org.truffleruby.core.thread;
+
+import org.truffleruby.builtins.CoreClass;
+import org.truffleruby.builtins.CoreMethod;
+import org.truffleruby.builtins.CoreMethodArrayArgumentsNode;
+import org.truffleruby.core.binding.BindingNodes;
+import org.truffleruby.language.arguments.RubyArguments;
+import org.truffleruby.language.methods.InternalMethod;
+
+import com.oracle.truffle.api.Truffle;
+import com.oracle.truffle.api.CompilerDirectives.TruffleBoundary;
+import com.oracle.truffle.api.dsl.Specialization;
+import com.oracle.truffle.api.frame.Frame;
+import com.oracle.truffle.api.frame.FrameInstance;
+import com.oracle.truffle.api.frame.FrameInstanceVisitor;
+import com.oracle.truffle.api.frame.MaterializedFrame;
+import com.oracle.truffle.api.object.DynamicObject;
+
+@CoreClass("Truffle::ThreadOperations")
+public class TruffleThreadNodes {
+
+    @CoreMethod(names = "ruby_caller", isModuleFunction = true, required = 0)
+    public abstract static class FindRubyCaller extends CoreMethodArrayArgumentsNode {
+
+        @Specialization
+        @TruffleBoundary
+        public DynamicObject executeFindRubyCaller(int skip) {
+            Frame rubyCaller = getFrame(2 + skip);
+            return rubyCaller == null ? nil() : BindingNodes.createBinding(getContext(), rubyCaller.materialize());
+        }
+
+        private MaterializedFrame getFrame(int skip) {
+            final Object frame = Truffle.getRuntime().iterateFrames(new FrameInstanceVisitor<Object>() {
+                int skipped = 0;
+
+                @Override
+                public Object visitFrame(FrameInstance frameInstance) {
+                    InternalMethod method = getMethod(frameInstance);
+                    if (method == null) {
+                        return null;
+                    } else if (!getContext().getCoreLibrary().isSend(method)) {
+                        if (skipped >= skip) {
+                            return frameInstance.getFrame(FrameInstance.FrameAccess.MATERIALIZE).materialize();
+                        } else {
+                            skipped++;
+                            return null;
+                        }
+                    } else {
+                        return null;
+                    }
```

Comment:
Could the CallStackManager be used for this logic, or this logic be moved there?
The method could just return a `FrameInstance` so it's independent of how it's used.